src/import/kv_importer.rs

```diff
@@ -0,0 +1,340 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs;
+use std::fmt;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use uuid::Uuid;
+use kvproto::importpb::*;
+
+use config::DbConfig;
+use util::collections::HashMap;
+
+use super::engine::*;
+use super::{Config, Error, Result};
+
+pub type Token = usize;
+
+struct Inner {
+    engines: HashMap<Uuid, Arc<EngineFile>>,
+    clients: HashMap<Token, Arc<EngineFile>>,
+}
+
+pub struct KVImporter {
+    dir: EngineDir,
+    token: AtomicUsize,
+    inner: Mutex<Inner>,
+}
+
+impl KVImporter {
+    pub fn new(cfg: Config, opts: DbConfig) -> Result<KVImporter> {
+        let dir = EngineDir::new(&cfg.import_dir, opts)?;
+        Ok(KVImporter {
+            dir: dir,
+            token: AtomicUsize::new(1),
+            inner: Mutex::new(Inner {
+                engines: HashMap::default(),
+                clients: HashMap::default(),
+            }),
+        })
+    }
+
+    pub fn token(&self) -> Token {
+        self.token.fetch_add(1, Ordering::SeqCst)
+    }
+
+    /// Open the engine.
+    pub fn open_engine(&self, uuid: Uuid) -> Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        if inner.engines.contains_key(&uuid) {
+            return Ok(());
+        }
+
+        match self.dir.open(uuid) {
+            Ok(engine) => {
+                info!("open {:?}", engine);
+                inner.engines.insert(uuid, Arc::new(engine));
+                Ok(())
+            }
+            Err(e) => {
+                error!("open {}: {:?}", uuid, e);
+                Err(e)
+            }
+        }
+    }
+
+    /// Write to the attached engine.
+    pub fn write_engine(&self, token: Token, batch: WriteBatch) -> Result<usize> {
```

Comment:
How about `get_engine`? So we can avoid lock-removal and lock-insertion over and over again.

Reply:
Any hint about how to do that? I have no idea how to carry the engine outside.

Reply:
Basically, I think it should look like:

```rust
let engine = {
  let mut inner = self.inner.lock().unwrap();
  inner.clients.get(&token).clone()
}

stream.fold(engine, |engine| { ... });
```

Reply:
Thanks for your advice, the code looks simpler now, PTAL.